src/adaptors/mod.rs

```diff
@@ -1117,6 +1117,73 @@ impl<I, F, T, U, E> Iterator for MapResults<I, F>
     }
 }
 
+/// An iterator adapter to filter values within a nested `Result`.
+///
+/// See [`.filter_results()`](../trait.Itertools.html#method.filter_results) for more information.
+#[must_use = "iterator adaptors are lazy and do nothing unless consumed"]
+pub struct FilterResults<I, F> {
+    iter: I,
+    f: F
+}
+
+/// Create a new `FilterResults` iterator.
+pub fn filter_results<I, F, T, E>(iter: I, f: F) -> FilterResults<I, F>
+    where I: Iterator<Item = Result<T, E>>,
+          F: FnMut(&T) -> bool,
+{
+    FilterResults {
+        iter: iter,
+        f: f,
+    }
+}
+
+impl<I, F, T, E> Iterator for FilterResults<I, F>
+    where I: Iterator<Item = Result<T, E>>,
+          F: FnMut(&T) -> bool,
+{
+    type Item = Result<T, E>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            match self.iter.next() {
+                Some(Ok(v)) => {
+                    if (self.f)(&v) {
+                        return Some(Ok(v));
+                    }
+                },
+                Some(Err(e)) => return Some(Err(e)),
+                None => return None,
+            }
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.iter.size_hint()
```

Comment:
I am not sure if this is correct. If `self.iter` iterates over e.g. a `Vec`, I assume that `self.iter.size_hint` returns `vec.len(), Some(vec.len())`, but due to filtering, our result should be `0, Some(vec.len())`.

Reply:
You assessment is correct. If this implementation will be used I will fix that.

Reply:
Can you fix this?